docs/design/1443-dynamic-pipelines.md

```diff
@@ -0,0 +1,252 @@
+## Abstract
+
+We are proposing a new feature to the [Alloy standard library][stdlib].
+It will be similar to a `map` operation over a collection such as a `list()`.
+Each `map` transformation will be done by a chain of components (a "sub-pipeline") created for this transformation.
+Each item in the collection will be processed by a different "sub-pipeline".
+
+The final solution may differ from a standard `map` operation, since there may be multiple outputs for the same input.
+For example, the sub-pipeline may branch into different `prometheus.relabel` components,
+each of which sends outputs to different components outside of the sub-pipeline.
+
+[stdlib]: https://grafana.com/docs/alloy/latest/reference/stdlib/
+
+## Use cases
+
+<!-- TODO: Add more use cases. It'd be helpful to gather feedback from the community and from solutions engineers. -->
+
+### Using discovery components together with prometheus.exporter ones
+
+Discovery components output a list of targets. It's not possible to input those lists directly to most exporter components.
+
+Suppose we have a list of targets produced by a `discovery` component:
+
+```
+[
+  {"__address__" = "redis-one:9115", "instance" = "one"},
+  {"__address__" = "redis-two:9116", "instance" = "two"},
+]
+```
+
+The [Alloy type][alloy-types] of the list above is `list(map(string))`.
+However, you may want to pipe information from this list of targets to a component which doesn't work with a `list()` or a `map()`.
+For example, you may want to input the `"__address__"` string to a `prometheus.exporter.redis`,
+and you may want to use the `"instance"` string in a `discovery.relabel`.
+
+[alloy-types]: https://grafana.com/docs/alloy/latest/get-started/configuration-syntax/expressions/types_and_values/
+
+## Proposal 1: A foreach block
+
+A `foreach` block may start several sub-pipelines for a `collection` specified in its arguments.
+
+```alloy
+// All components in the sub-pipeline will be scoped under "foreach.default/1/...".
+// Here, "1" is sub-pipeline number 1.
+// This way component names won't clash with other sub-pipelines from the same foreach, 
+// and with the names of components outside of the foreach.
+foreach "default" {
+    
+  // "collection" is what the for loop will iterate over.
+  collection = discovery.file.default.targets
+
+  // Each item in the collection will be accessible via the "target" variable.
+  // E.g. `target["__address__"]`.
+  var = "target"
+
+  // A sub-pipeline consisting of components which process each target.
+  ...
+}
+```
+
+<details>
+  <summary>Example</summary>
+
+```alloy
+discovery.file "default" {
+  files = ["/Users/batman/Desktop/redis_addresses.yaml"]
+}
+
+// Every component defined in the "foreach" block will be instantiated for each item in the collection. 
+// The instantiated components will be scoped using the name of the foreach block and the index of the
+// item in the collection. For example: /foreach.redis/0/prometheus.exporter.redis.default
```

Comment:
> /foreach.redis/0/prometheus.exporter.redis.default

Targets can come and go between discovery intervals. If you use the index to scope them you will need to shift it on every update. Maybe you can just use the target itself which should be unique: /foreach.redis/target/prometheus.exporter.redis.default

Reply:
> Maybe you can just use the target itself which should be unique

What do you mean by "the target itself"? Do you mean the "__address__" label? We're not guaranteed to have such a label on the collection which is being iterated on :) The collection might not be `targets`. Also, some `targets` might not follow this convention of having an `__address__` label.

Reply:
the target (or more generally the item in the collection) can be hashed and we can use the hash for scoping and for avoid recomputations on update when a new collection comes

Reply:
Yeah, I think targets going away in the middle of the list is a good point to consider. Hashing the way that William is proposing might work and we'd need to recommend people to try change the input targets as little as possible (e.g. use a reasonable refresh interval), otherwise there will be a lot of churn.

Reply:
Theoretically, couldn't there be hash collisions? I don't see how hashing is a very resilient way of handling this.

Reply:
With [sufficiently long hash](https://stackoverflow.com/questions/62664761/probability-of-hash-collision) you don't need to worry about collisions. Unless there are implementation bugs. 

If collisions worry you or we want to use shorter IDs than a 512bit hash, we could think of a mapper that will assign stable IDs to targets, using a simple `int` sequence.